### PR TITLE
Wrong token count in validation when CP>1

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
## Wrong token count in validation

When `CP > 1`, `post_dataloading_process` shards inputs across ranks. During validation, this causes token counts to be incorrect because counting happens after sharding. This is not an issue during training, where tokens are counted before `post_dataloading_process`.

### Fix

Move `post_dataloading_process` to occur after token counting in validation.

## `post_dataloading_process` not receiving the correct `attn_config`

At some point, the validation and trainer implementations of `post_dataloading_process` became misaligned. In validation, `attn_config` was not being retrieved correctly and defaulted to `None`. As a consequence, the mask type defaulted to `causal` even when `block_causal` was configured.

### Fix

As a temporary fix, I copied the relevant logic from the trainer into the validation `post_dataloading_process`.

This comment highlights the issue:
https://github.com/pytorch/torchtitan/blob/ae5ff104999a937ce0d34d8a720969ae23110140/torchtitan/components/validate.py#L181-L185

I can open a follow-up PR to address this more cleanly. A few possible approaches:

* Make `Validator` inherit from the trainer so it reuses `post_dataloading_process` directly. This feels somewhat heavy-handed since that is the only functionality we need to share.
* Pass `post_dataloading_process` into `Validator.__init__`. We could also make the trainer implementation a `@staticmethod` to reduce coupling.

### CODE REVIEW

File: torchtitan/components/validate.py [hunk 1]
```diff
@@ -184,30 +184,31 @@ def post_dataloading_process(
...
+                positions = torch.arange(
+                    inputs.shape[1], dtype=torch.int32, device=inputs.device
+                ).repeat(inputs.shape[0], 1)
 
         try:
-            # pyrefly: ignore [not-callable]
             extra_kwargs["attention_masks"] = cast(
-                BaseModel, model_parts[0]
+                Decoder, model_parts[0]
             ).get_attention_masks(
```
Comment (wwwjn) [inline torchtitan/components/validate.py]: Why we don't put get_attention_mask() under `if isinstance(model_config, Decoder.Config):`

Mirroring: https://github.com/pytorch/torchtitan/blob/refs/heads/main/torchtitan/trainer.py#L611

**Review by wwwjn — COMMENTED** [review]
Nice catch, thanks! Can you attach an actual run log to show grad_norm difference before v.s. after this change?

In long term, validate.py shares a lot of duplicate code with trainer.py

File: torchtitan/components/validate.py (hunk 1 above)
Comment (francesco-bertolotti) [reply torchtitan/components/validate.py]: Good point, fixing it now!

**Comment by francesco-bertolotti** [comment]
Hi @wwwjn,

This bug does not affect training behavior in any way. The issue was only in validation reporting: the validation loss appeared roughly 4X (bc I am running with 4 ranks) higher than it actually was because `global_valid_tokens` was miscalculated. There is no impact on gradient norms or optimization.

Here are two comparison runs:

This is the validation loss before (`test`) and after (`fixed`) my commits: valid

This is the gradient norm before and after the commits (unchanged): grads

Configuration used:

```python
torchtitan.trainer.Trainer.Config(
    model_spec=torchtitan.models.llama3.model_registry("debugmodel", attn_backend="flex"),
    dataloader = torchtitan.hf_datasets.text_datasets.HuggingFaceTextDataLoader.Config(
        dataset = "c4_test",
        num_workers = 1,
    ),
    parallelism = torchtitan.trainer.ParallelismConfig(
        data_parallel_replicate_degree=1, 
        data_parallel_shard_degree=-1,
        context_parallel_degree=4,
        context_parallel_load_balancer=None,
    ),
    compile = torchtitan.trainer.CompileConfig(
        enable=False,
    ),
    optimizer = torchtitan.components.optimizer.OptimizersContainer.Config(
        lr=1e-4,
    ),
    training = torchtitan.config.TrainingConfig(
        local_batch_size=1,
        seq_len=512,
        steps=100,
        dtype="float32",
        mixed_precision_param="float32"
    ),
    loss = torchtitan.components.loss.CrossEntropyLoss.Config(),
    lr_scheduler=torchtitan.components.lr_scheduler.LRSchedulersContainer.Config(
        warmup_steps=10,
        decay_ratio=0,
        decay_type="linear",
        min_lr_factor=0.0,
    ),
    checkpoint=torchtitan.components.checkpoint.CheckpointManager.Config(
        interval=1000,
        last_save_model_only=False,
        export_dtype="bfloat16",
        enable=True,
    ),
    activation_checkpoint=torchtitan.config.ActivationCheckpointConfig(
        mode="selective",
    ),
    metrics=torchtitan.components.metrics.MetricsProcessor.Config(
        enable_tensorboard=True,
        enable_wandb=False,
        log_freq=1
    ),
    debug=torchtitan.config.configs.DebugConfig(
        seed=42,
        deterministic=True,
    ),
    validator=torchtitan.components.validate.Validator.Config(
        enable=True,
        freq=10,
        dataloader=torchtitan.hf_datasets.text_datasets.HuggingFaceTextDataLoader.Config(
            dataset = "c4_test",
            dataset_path = None,
            infinite = False,
        ),
        steps=3,
    ),
    hf_assets_path="./tests/assets/tokenizer",
    dump_folder=".data/experiments/test",
)
```

File: torchtitan/components/validate.py [hunk 2]
```diff
@@ -184,29 +185,30 @@ def post_dataloading_process(
...
-        try:
-            # pyrefly: ignore [not-callable]
-            extra_kwargs["attention_masks"] = cast(
-                BaseModel, model_parts[0]
-            ).get_attention_masks(
-                positions=extra_kwargs.get("positions"),
-            )
-        except TypeError:
-            pass
+        if isinstance(model_config, Decoder.Config):
```
Comment (fegin) [inline torchtitan/components/validate.py:190]: Is this a logic change or simply a typing improvement? I think it is the later but I want to confirm.

**Review by tianyu-l — APPROVED** [review]
oh, still an issue caused by validator not in sync with trainer.
In the long term, we should think about what refactor makes the most sense.
In the short term, we can add validator loss to be part of the CI loss guard. cc @wwwjn

File: torchtitan/components/validate.py (hunk 2 above)
Comment (francesco-bertolotti) [reply torchtitan/components/validate.py:190]: I am not sure either. I just aligned the code with the one from the trainer:
https://github.com/pytorch/torchtitan/blob/8599818f813ae43d73aeb142a879cbbff9b1d9ad/torchtitan/trainer.py#L594-L614

If this is a problem, I can change it/

Comment (fegin) [reply torchtitan/components/validate.py:190]: That's fine. It is a typing fix. Thanks!

**Review by fegin — APPROVED** [review]
